### PR TITLE
ECC and TLS support for all SECP, Koblitz and Brainpool curves

### PR DESCRIPTION
**Pull request by dgarske**
Added ECC and TLS support for all SECP, Koblitz and Brainpool curves. Adds new "HAVE_ECC_BRAINPOOL", "HAVE_ECC_KOBLITZ", "HAVE_ECC_SECPR2" and "HAVE_ECC_SECPR3" options. ECC refactor to use curve_id in _ex functions. NID and ECC Id's match now. Added ability to encode OID (HAVE_OID_ENCODING), but leave off by default and will use pre-encoded value for best performance.

To enable all features for testing use:
./configure --enable-keygen --enable-certgen --enable-opensslextra --enable-supportedcurves CFLAGS="-DHAVE_ECC_BRAINPOOL -DHAVE_ECC_KOBLITZ -DHAVE_ECC_SECPR2 -DHAVE_ECC_SECPR3 -DWOLFSSL_CUSTOM_CURVES -DHAVE_OID_ENCODING -DECC_DUMP_OID"

### CODE REVIEW

File: wolfcrypt/src/ecc.c [hunk 1]
```diff
@@ -41,29 +41,51 @@ Possible ECC enable options:
...
  * WOLFSSL_CUSTOM_CURVES: Allow non-standard curves.            default: off
- *                         Includes the curve "a" variable in calculation
+ *                        Includes the curve "a" variable in calculation
+ * ECC_DUMP_OID:        Enables dump of OID encoding and sum   default: off
  */
 
 /*
-ECC Curves:
+ECC Curve Types:
+ * NO_ECC_SECP          Disables SECP curves                    default: on
```
Comment (toddouska) [inline wolfcrypt/src/ecc.c]: Isn't the default for NO_ECC_SECP off?  That is, ECC_SECP is on.

Comment (dgarske) [inline wolfcrypt/src/ecc.c]: Yeah its not defined by default, which allows SECP to be used. I'll update that lingo. Thx

**Comment by toddouska** [comment]
Feature request:  Detect a == 3 case for prime curves and use optimized code when CUSTOM_CURVES is on.

**Comment by toddouska** [comment]
Thanks, merging.  Will need an additional PR for doc fix and a == -3 SECP optimization.

**Pull request merged**